{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Recuperación de las secuencias proteicas virales\n",
    "Las secuencias proteicas de las <u>10 poliproteínas virales del DENV-2</u> se recuperaron de la base de datos del '**Universal Protein Resource**' - *Uniprot* (https://www.uniprot.org/) (*Uniprot Consortium, 2023*), a partir de una poliproteína de *3 391 aminoácidos* del genoma de DENV-2 (código de acceso: __Q9WDA6__) procedente de Iquitos y colectada en 1996, la cual es la única poliproteína en Perú que ha sido revisada en Uniprot hasta el momento, formando parte de la sección '*UniProtKB/Swiss-Prot*'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting Pillow\n",
      "  Downloading Pillow-9.5.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m5.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: Pillow\n",
      "Successfully installed Pillow-9.5.0\n"
     ]
    }
   ],
   "source": [
    "# Instalar el paquete 'Pillow'\n",
    "!python3 -m pip install --upgrade Pillow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting bioservices\n",
      "  Using cached bioservices-1.11.2.tar.gz (191 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting appdirs (from bioservices)\n",
      "  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
      "Collecting beautifulsoup4 (from bioservices)\n",
      "  Using cached beautifulsoup4-4.12.2-py3-none-any.whl (142 kB)\n",
      "Collecting colorlog (from bioservices)\n",
      "  Using cached colorlog-6.7.0-py2.py3-none-any.whl (11 kB)\n",
      "Collecting easydev>=0.12.0 (from bioservices)\n",
      "  Using cached easydev-0.12.1.tar.gz (55 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting grequests (from bioservices)\n",
      "  Using cached grequests-0.6.0-py3-none-any.whl (5.2 kB)\n",
      "Collecting matplotlib (from bioservices)\n",
      "  Using cached matplotlib-3.5.3-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (11.2 MB)\n",
      "Collecting pandas>=0.21.1 (from bioservices)\n",
      "  Downloading pandas-1.3.5-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.3 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.3/11.3 MB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hCollecting requests (from bioservices)\n",
      "  Using cached requests-2.31.0-py3-none-any.whl (62 kB)\n",
      "Collecting requests_cache (from bioservices)\n",
      "  Using cached requests_cache-1.0.1-py3-none-any.whl (58 kB)\n",
      "Collecting suds-community>=0.7 (from bioservices)\n",
      "  Using cached suds_community-1.1.2-py3-none-any.whl (144 kB)\n",
      "Collecting lxml (from bioservices)\n",
      "  Downloading lxml-4.9.2-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (6.6 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.6/6.6 MB\u001b[0m \u001b[31m5.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting tqdm (from bioservices)\n",
      "  Using cached tqdm-4.65.0-py3-none-any.whl (77 kB)\n",
      "Collecting wrapt (from bioservices)\n",
      "  Using cached wrapt-1.15.0-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (75 kB)\n",
      "Collecting xmltodict (from bioservices)\n",
      "  Using cached xmltodict-0.13.0-py2.py3-none-any.whl (10.0 kB)\n",
      "Collecting colorama (from easydev>=0.12.0->bioservices)\n",
      "  Using cached colorama-0.4.6-py2.py3-none-any.whl (25 kB)\n",
      "Requirement already satisfied: pexpect in /snap/jupyter/6/lib/python3.7/site-packages (from easydev>=0.12.0->bioservices) (4.7.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /snap/jupyter/6/lib/python3.7/site-packages (from pandas>=0.21.1->bioservices) (2.8.0)\n",
      "Collecting pytz>=2017.3 (from pandas>=0.21.1->bioservices)\n",
      "  Using cached pytz-2023.3-py2.py3-none-any.whl (502 kB)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /home/victor/snap/jupyter/common/lib/python3.7/site-packages (from pandas>=0.21.1->bioservices) (1.21.6)\n",
      "Collecting soupsieve>1.2 (from beautifulsoup4->bioservices)\n",
      "  Using cached soupsieve-2.4.1-py3-none-any.whl (36 kB)\n",
      "Collecting gevent (from grequests->bioservices)\n",
      "  Downloading gevent-22.10.2-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.0 MB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.0/6.0 MB\u001b[0m \u001b[31m3.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0mm eta \u001b[36m0:00:01\u001b[0m0:01\u001b[0m:01\u001b[0m\n",
      "\u001b[?25hCollecting cycler>=0.10 (from matplotlib->bioservices)\n",
      "  Using cached cycler-0.11.0-py3-none-any.whl (6.4 kB)\n",
      "Collecting fonttools>=4.22.0 (from matplotlib->bioservices)\n",
      "  Using cached fonttools-4.38.0-py3-none-any.whl (965 kB)\n",
      "Collecting kiwisolver>=1.0.1 (from matplotlib->bioservices)\n",
      "  Using cached kiwisolver-1.4.4-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.1 MB)\n",
      "Collecting packaging>=20.0 (from matplotlib->bioservices)\n",
      "  Using cached packaging-23.1-py3-none-any.whl (48 kB)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/victor/snap/jupyter/common/lib/python3.7/site-packages (from matplotlib->bioservices) (9.5.0)\n",
      "Collecting pyparsing>=2.2.1 (from matplotlib->bioservices)\n",
      "  Using cached pyparsing-3.0.9-py3-none-any.whl (98 kB)\n",
      "Collecting charset-normalizer<4,>=2 (from requests->bioservices)\n",
      "  Downloading charset_normalizer-3.1.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (171 kB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m171.0/171.0 kB\u001b[0m \u001b[31m6.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting idna<4,>=2.5 (from requests->bioservices)\n",
      "  Using cached idna-3.4-py3-none-any.whl (61 kB)\n",
      "Collecting urllib3<3,>=1.21.1 (from requests->bioservices)\n",
      "  Using cached urllib3-2.0.2-py3-none-any.whl (123 kB)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /snap/jupyter/6/lib/python3.7/site-packages (from requests->bioservices) (2019.3.9)\n",
      "Collecting attrs>=21.2 (from requests_cache->bioservices)\n",
      "  Using cached attrs-23.1.0-py3-none-any.whl (61 kB)\n",
      "Collecting cattrs>=22.2 (from requests_cache->bioservices)\n",
      "  Using cached cattrs-22.2.0-py3-none-any.whl (35 kB)\n",
      "Collecting platformdirs>=2.5 (from requests_cache->bioservices)\n",
      "  Using cached platformdirs-3.5.1-py3-none-any.whl (15 kB)\n",
      "Collecting url-normalize>=1.4 (from requests_cache->bioservices)\n",
      "  Using cached url_normalize-1.4.3-py2.py3-none-any.whl (6.8 kB)\n",
      "Collecting importlib-metadata (from attrs>=21.2->requests_cache->bioservices)\n",
      "  Using cached importlib_metadata-6.6.0-py3-none-any.whl (22 kB)\n",
      "Collecting exceptiongroup (from cattrs>=22.2->requests_cache->bioservices)\n",
      "  Using cached exceptiongroup-1.1.1-py3-none-any.whl (14 kB)\n",
      "Collecting typing_extensions (from cattrs>=22.2->requests_cache->bioservices)\n",
      "  Using cached typing_extensions-4.6.2-py3-none-any.whl (31 kB)\n",
      "Requirement already satisfied: six>=1.5 in /snap/jupyter/6/lib/python3.7/site-packages (from python-dateutil>=2.7.3->pandas>=0.21.1->bioservices) (1.12.0)\n",
      "Collecting zope.event (from gevent->grequests->bioservices)\n",
      "  Using cached zope.event-4.6-py2.py3-none-any.whl (6.8 kB)\n",
      "Collecting zope.interface (from gevent->grequests->bioservices)\n",
      "  Using cached zope.interface-6.0-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (241 kB)\n",
      "Requirement already satisfied: setuptools in /snap/jupyter/6/lib/python3.7/site-packages (from gevent->grequests->bioservices) (41.0.1)\n",
      "Collecting greenlet>=2.0.0 (from gevent->grequests->bioservices)\n",
      "  Downloading greenlet-2.0.2-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (566 kB)\n",
      "\u001b[2K     \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m566.1/566.1 kB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m[31m7.6 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: ptyprocess>=0.5 in /snap/jupyter/6/lib/python3.7/site-packages (from pexpect->easydev>=0.12.0->bioservices) (0.6.0)\n",
      "Collecting zipp>=0.5 (from importlib-metadata->attrs>=21.2->requests_cache->bioservices)\n",
      "  Using cached zipp-3.15.0-py3-none-any.whl (6.8 kB)\n",
      "Building wheels for collected packages: bioservices, easydev\n",
      "  Building wheel for bioservices (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for bioservices: filename=bioservices-1.11.2-py3-none-any.whl size=223226 sha256=08d7e626155fed581e265292e910f22c7a06bb26433187cb9ded28faa44ddb0f\n",
      "  Stored in directory: /home/victor/snap/jupyter/6/.cache/pip/wheels/d8/d0/1b/096c8136d4279f009a78ee11e12630a652f4351e70b2765365\n",
      "  Building wheel for easydev (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for easydev: filename=easydev-0.12.1-py3-none-any.whl size=64201 sha256=4ebebaa8472436b5caf410bab0307c31b05d919d1bf0a5d19abec330ca86a58c\n",
      "  Stored in directory: /home/victor/snap/jupyter/6/.cache/pip/wheels/69/c0/46/1e37597fd8142973e0bc798c8c217f9d91940c21b4e39e29cb\n",
      "Successfully built bioservices easydev\n",
      "Installing collected packages: pytz, appdirs, zope.interface, zope.event, zipp, xmltodict, wrapt, urllib3, url-normalize, typing_extensions, tqdm, suds-community, soupsieve, pyparsing, packaging, lxml, idna, greenlet, fonttools, exceptiongroup, cycler, colorlog, colorama, charset-normalizer, requests, platformdirs, pandas, kiwisolver, importlib-metadata, gevent, easydev, beautifulsoup4, matplotlib, grequests, attrs, cattrs, requests_cache, bioservices\n",
      "Successfully installed appdirs-1.4.4 attrs-23.1.0 beautifulsoup4-4.12.2 bioservices-1.11.2 cattrs-22.2.0 charset-normalizer-3.1.0 colorama-0.4.6 colorlog-6.7.0 cycler-0.11.0 easydev-0.12.1 exceptiongroup-1.1.1 fonttools-4.38.0 gevent-22.10.2 greenlet-2.0.2 grequests-0.6.0 idna-3.4 importlib-metadata-6.6.0 kiwisolver-1.4.4 lxml-4.9.2 matplotlib-3.5.3 packaging-23.1 pandas-1.3.5 platformdirs-3.5.1 pyparsing-3.0.9 pytz-2023.3 requests-2.31.0 requests_cache-1.0.1 soupsieve-2.4.1 suds-community-1.1.2 tqdm-4.65.0 typing_extensions-4.6.2 url-normalize-1.4.3 urllib3-2.0.2 wrapt-1.15.0 xmltodict-0.13.0 zipp-3.15.0 zope.event-4.6 zope.interface-6.0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "# Instalar el paquete 'bioservices'\n",
    "%pip install bioservices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importar paquete de librería estándar (io), el módulo de 'bioservices' y 'Pandas'\n",
    "import io\n",
    "from bioservices import UniProt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating directory /home/victor/snap/jupyter/6/.cache/bioservices \n",
      "Welcome to Bioservices\n",
      "======================\n",
      "It looks like you do not have a configuration file.\n",
      "We are creating one with default values in /home/victor/snap/jupyter/6/.config/bioservices/bioservices.cfg .\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "# Conectar a UniProt\n",
    "service = UniProt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entry\tEntry Name\tReviewed\tProtein names\tGene Names\tOrganism\tLength\n",
      "Q9WDA6\tPOLG_DEN2Q\treviewed\tGenome polyprotein [Cleaved into: Capsid protein C (Core protein); Protein prM; Peptide pr; Small envelope protein M (Matrix protein); Envelope protein E; Non-structural protein 1 (NS1); Non-structural protein 2A (NS2A); Serine protease subunit NS2B (Flavivirin protease NS2B regulatory subunit) (Non-structural protein 2B); Serine protease NS3 (EC 3.4.21.91) (EC 3.6.1.15) (EC 3.6.4.13) (Flavivirin protease NS3 catalytic subunit) (Non-structural protein 3); Non-structural protein 4A (NS4A); Peptide 2k; Non-structural protein 4B (NS4B); RNA-directed RNA polymerase NS5 (EC 2.1.1.56) (EC 2.1.1.57) (EC 2.7.7.48) (Non-structural protein 5)]\t\tDengue virus type 2 (strain Peru/IQT2913/1996) (DENV-2)\t3391\n"
     ]
    }
   ],
   "source": [
    "# Construir y realizar la consulta\n",
    "query = 'Q9WDA6'\n",
    "result = service.search(query)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9WDA6</td>\n",
       "      <td>POLG_DEN2Q</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Genome polyprotein [Cleaved into: Capsid prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dengue virus type 2 (strain Peru/IQT2913/1996)...</td>\n",
       "      <td>3391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry  Entry Name  Reviewed  \\\n",
       "0  Q9WDA6  POLG_DEN2Q  reviewed   \n",
       "\n",
       "                                       Protein names  Gene Names  \\\n",
       "0  Genome polyprotein [Cleaved into: Capsid prote...         NaN   \n",
       "\n",
       "                                            Organism  Length  \n",
       "0  Dengue virus type 2 (strain Peru/IQT2913/1996)...    3391  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##  Convertir el último resultado de búsqueda en un marco de datos en Pandas\n",
    "df_table = pd.read_table(io.StringIO(result))\n",
    "df_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">sp|Q9WDA6|POLG_DEN2Q Genome polyprotein OS=Dengue virus type 2 (strain Peru/IQT2913/1996) OX=408694 PE=1 SV=1\n",
      "MNNQRKKARNTPFNMLKRERNRVSTVQQLTKRFSLGMLQGRGPLKLFMALVAFLRFLTIP\n",
      "PTAGILKRWGTIKKSKAINVLRGFRKEIGRMLNILNRRRRTAGMIIMLIPTVMAFHLTTR\n",
      "NGEPHMIVSRQEKGKSLLFKTKDGTNMCTLMAMDLGELCEDTITYKCPFLKQNEPEDIDC\n",
      "WCNSTSTWVTYGTCTTTGEHRREKRSVALVPHVGMGLETRTETWMSSEGAWKHAQRIETW\n",
      "ILRHPGFTIMAAILAYTIGTTHFQRVLIFILLTAIAPSMTMRCIGISNRDFVEGVSGGSW\n",
      "VDIVLEHGSCVTTMAKNKPTLDFELIKTEAKQPATLRKYCIEAKLTNTTTDSRCPTQGEP\n",
      "TLNEEQDKRFVCKHSMVDRGWGNGCGLFGKGGIVTCAMFTCKKNMEGKIVQPENLEYTVV\n",
      "ITPHSGEEHAVGNDTGKHGKEVKITPQSSITEAELTGYGTVTMECSPRTGLDFNEMVLLQ\n",
      "MEDKAWLVHRQWFLDLPLPWLPGADTQGSNWIQKETLVTFKNPHAKKQDVVVLGSQEGAM\n",
      "HTALTGATEIQMSSGNLLFTGHLKCRLRMDKLQLKGMSYSMCTGKFKIVKEIAETQHGTI\n",
      "VIRVQYEGDGSPCKIPFEIMDLEKRHVLGRLITVNPIVTEKDSPVNIEAEPPFGDSYIII\n",
      "GAEPGQLKLDWFKKGSSIGQMFETTMRGAKRMAILGDTAWDFGSLGGVFTSIGKALHQVF\n",
      "GAIYGAAFSGVSWTMKILIGVIITWIGMNSRSTSLSVSLVLVGIVTLYLGVMVQADSGCV\n",
      "VSWKNKELKCGSGIFVTDNVHTWTEQYKFQPESPSKLASAIQKAHEEGICGIRSVTRLEN\n",
      "LMWKQITSELNHILSENEVKLTIMTGDIKGIMQVGKRSLRPQPTELRYSWKTWGKAKMLS\n",
      "TELHNQTFLIDGPETAECPNTNRAWNSLEVEDYGFGVFTTNIWLRLREKQDAFCDSKLMS\n",
      "AAIKDNRAVHADMGYWIESALNDTWKIEKASFIEVKSCHWPKSHTLWSNGVLESEMVIPK\n",
      "NIAGPVSQHNNRPGYHTQTAGPWHLGKLEMDFDFCEGTTVVVTEECGNRGPSLRTTTASG\n",
      "KLITEWCCRSCTLPPLRYRGEDGCWYGMEIRPLKEKEENLVSSLVTAGHGQIDNFSLGIL\n",
      "GMALFLEEMLRTRVGTKHAILLVAVSFLTLITGNMSFRDLGRVMVMVGATMTDDIGMGVT\n",
      "YLALLAAFKVRPTFAAGLLLRKLTSKELMMTTIGIVLLSQSSIPETILELTDALALGMMV\n",
      "LKMVRNMEKYQLAVTIMAILCVPNAVILQNAWKVSCTTLAVVSVSPLLLTSSQQKADWIP\n",
      "LALTIKGLNPTAIFLTTLTRTSKKRSWPLNEAIMAVGMVSILASSLLKNDIPMTGPLVAG\n",
      "GLLTVCYVLTGRSADLELERATDVKWDDQAEISGSSPILSITISEDGSMSIKNEEEEQTL\n",
      "TILIRTGLLVISGLFPVSIPITAAAWYLWEVKKQRAGVLWDVPSPPPVGRAELEDGAYRI\n",
      "KQKGILGYSQIGAGVYKEGTFHTMWHVTRGAVLMHKGKRIEPSWADVKKDLISYGGGWKL\n",
      "EGEWKEGEEVQVLALEPGKNPRAVQTKPGLFRTNTGTIGAVSLDFSPGTSGSPIVDKKGK\n",
      "VVGLYGNGVVTRGGAYVSAIAQTEKGIEDNPEIEDDIFRKRRLTIMDLHPGAGKTKRYLP\n",
      "AIVREAIKRGLRTLILAPTRVVAAEMEEALRGLPIRYQTPAIRAEHTGREIVDLMCHATF\n",
      "TMRLLSPIRVPNYNLIIMDEAHFTDPASIAARGYISTRVEMGEAAGIFMTATPPGSRDPF\n",
      "PQSNAPIMDEEREIPERSWNSGHEWVTDFKGKTVWFVPSIKTGNDIAACLRKNGKRVIQL\n",
      "SRKTFDSEYVKTRTNDWDFVVTTDISEMGANFKAERVIDPRRCMKPVILTDGEERVILAG\n",
      "PMPVTHSSAAQRRGRIGRNPRNENDQYIYMGEPLENDEDCAHWKEAKMLLDNINTPEGII\n",
      "PSMFEPEREKVDAIDGEYRLRGEARKTFVDLMRRGDLPVWLAYKVAAEGINYADRRWCFD\n",
      "GTRNNQILEENVEVEIWTKEGERKKLKPRWLDARIYSDPLALKEFKEFAAGRKSLTLNLI\n",
      "TEMGRLPTFMTQKARDALDNLAVLHTAEAGGKAYNHALSELPETLETLLLLTLLATVTGG\n",
      "IFLFLMSGRGIGKMTLGMCCIITASILLWYAQIQPHWIAASIILEFFLIVLLIPEPEKQR\n",
      "TPQDNQLTYVIIAILTVVAATMANEMGFLEKTKKDLGLGHIATQQPESNILDIDLRPASA\n",
      "WTLYAVATTFITPMLRHSIENSSVNVSLTAIANQATVLMGLGKGWPLSKMDIGVPLLAIG\n",
      "CYSQVNPITLTAALLMLVAHYAIIGPGLQAKATREAQKRAAAGIMKNPTVDGITVIDLDP\n",
      "IPYDPKFEKQLGQVMLLVLCVTQVLMMRTTWALCEALTLATGPVSTLWEGNPGRFWNTTI\n",
      "AVSMANIFRGSYLAGAGLLFSIMKNTTSTRRGTGNMGETLGEKWKNRLNALGKSEFQIYK\n",
      "KSGIQEVDRTLAKEGIKRGETDHHAVSRGSAKLRWFVERNLVTPEGKVVDLGCGRGGWSY\n",
      "YCGGLKNVREVKGLTKGGPGHEEPIPMSTYGWNLVRLQSGVDVFFVPPEKCDTLLCDIGE\n",
      "SSPNPTVEAGRTLRVLNLVENWLNNNTQFCVKVLNPYMPSVIERMETLQRKYGGALVRNP\n",
      "LSRNSTHEMYWVSNASGNIVSSVNMISRMLINRFTMRHKKATYEPDVDLGSGTRNIGIES\n",
      "ETPNLDIIGKRIEKIKQEHETSWHYDQDHPYKTWAYHGSYETKQTGSASSMVNGVVRLLT\n",
      "KPWDVIPMVTQMAMTDTTPFGQQRVFKEKVDTRTQEPKEGTKKLMKITAEWLWKELGKKK\n",
      "TPRMCTREEFTKKVRSNAALGAIFTDENKWKSAREAVEDNRFWELVDKERNLHLEGKCET\n",
      "CVYNMMGKREKKLGEFGKAKGSRAIWYMWLGARFLEFEALGFLNEDHWFSRENSLSGVEG\n",
      "EGLHKLGYILREVSKKEGGAMYADDTAGWDTRITIEDLKNEEMITNHMAGEHKKLAEAIF\n",
      "KLTYQNKVVRVQRPTPRGTVMDIISRRDQRGSGQVVTYGLNTFTNMEAQLIRQMEGEGVF\n",
      "KSIQHLTASEEIAVQDWLVRVGRERLSRMAISGDDCVVKPLDDRFAKALTALNDMGKVRK\n",
      "DIQQWEPSRGWNDWTQVPFCSHHFHELIMKDGRTLVVPCRNQDELIGRARISQGAGWSLR\n",
      "ETACLGKSYAQMWSLMYFHRRDLRLAANAICSAVPSHWVPTSRTTWSIHASHEWMTTEDM\n",
      "LTVWNRVWILENPWMEDKTPVESWEEIPYLGKREDQWCGSLIGLTSRATWAKNIQTAINQ\n",
      "VRSLIGNEEYTDYMPSMKRFRREEEEVGVLW\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Visualizar la secuencia aminoacídica completa del genoma de la poliproteína del DENV-2\n",
    "result_fasta = service.search(query, frmt = 'fasta')\n",
    "print(result_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Guardar la secuencia aminoacídica completa del genoma de la poliproteína del DENV-2 (Formato fasta)\n",
    "result_fasta = service.search(query, frmt = 'fasta')\n",
    "file_genoma = 'Q9WDA6 - DENV_2 - genome polyprotein'\n",
    "\n",
    "with open(file_genoma, 'w') as file:\n",
    "    # Write the content to the file\n",
    "    file.write(result_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">ancC_Proteína de la cápside anclada\n",
      "MNNQRKKARNTPFNMLKRERNRVSTVQQLTKRFSLGMLQGRGPLKLFMALVAFLRFLTIP\n",
      "PTAGILKRWGTIKKSKAINVLRGFRKEIGRMLNILNRRRR\n",
      "\n",
      ">prM_Proteína premembrana\n",
      "FHLTTR\n",
      "NGEPHMIVSRQEKGKSLLFKTKDGTNMCTLMAMDLGELCEDTITYKCPFLKQNEPEDIDC\n",
      "WCNSTSTWVTYGTCTTTGEHRREKRSVALVPHVGMGLETRTETWMSSEGAWKHAQRIETW\n",
      "ILRHPGFTIMAAILAYTIGTTHFQRVLIFILLTAIAPSMT\n",
      "\n",
      ">E_Proteína de envoltura\n",
      "MRCIGISNRDFVEGVSGGSW\n",
      "VDIVLEHGSCVTTMAKNKPTLDFELIKTEAKQPATLRKYCIEAKLTNTTTDSRCPTQGEP\n",
      "TLNEEQDKRFVCKHSMVDRGWGNGCGLFGKGGIVTCAMFTCKKNMEGKIVQPENLEYTVV\n",
      "ITPHSGEEHAVGNDTGKHGKEVKITPQSSITEAELTGYGTVTMECSPRTGLDFNEMVLLQ\n",
      "MEDKAWLVHRQWFLDLPLPWLPGADTQGSNWIQKETLVTFKNPHAKKQDVVVLGSQEGAM\n",
      "HTALTGATEIQMSSGNLLFTGHLKCRLRMDKLQLKGMSYSMCTGKFKIVKEIAETQHGTI\n",
      "VIRVQYEGDGSPCKIPFEIMDLEKRHVLGRLITVNPIVTEKDSPVNIEAEPPFGDSYIII\n",
      "GAEPGQLKLDWFKKGSSIGQMFETTMRGAKRMAILGDTAWDFGSLGGVFTSIGKALHQVF\n",
      "GAIYGAAFSGVSWTMKILIGVIITWIGMNSRSTSLSVSLVLVGIVTLYLGVMVQA\n",
      "\n",
      ">NS1_Proteína no estructural 1\n",
      "DSGCV\n",
      "VSWKNKELKCGSGIFVTDNVHTWTEQYKFQPESPSKLASAIQKAHEEGICGIRSVTRLEN\n",
      "LMWKQITSELNHILSENEVKLTIMTGDIKGIMQVGKRSLRPQPTELRYSWKTWGKAKMLS\n",
      "TELHNQTFLIDGPETAECPNTNRAWNSLEVEDYGFGVFTTNIWLRLREKQDAFCDSKLMS\n",
      "AAIKDNRAVHADMGYWIESALNDTWKIEKASFIEVKSCHWPKSHTLWSNGVLESEMVIPK\n",
      "NIAGPVSQHNNRPGYHTQTAGPWHLGKLEMDFDFCEGTTVVVTEECGNRGPSLRTTTASG\n",
      "KLITEWCCRSCTLPPLRYRGEDGCWYGMEIRPLKEKEENLVSSLVTA\n",
      "\n",
      ">NS2A_Proteína no estructural 2A\n",
      "GHGQIDNFSLGIL\n",
      "GMALFLEEMLRTRVGTKHAILLVAVSFLTLITGNMSFRDLGRVMVMVGATMTDDIGMGVT\n",
      "YLALLAAFKVRPTFAAGLLLRKLTSKELMMTTIGIVLLSQSSIPETILELTDALALGMMV\n",
      "LKMVRNMEKYQLAVTIMAILCVPNAVILQNAWKVSCTTLAVVSVSPLLLTSSQQKADWIP\n",
      "LALTIKGLNPTAIFLTTLTRTSKKR\n",
      "\n",
      ">NS2B_Proteína no estructural 2B\n",
      "SWPLNEAIMAVGMVSILASSLLKNDIPMTGPLVAG\n",
      "GLLTVCYVLTGRSADLELERATDVKWDDQAEISGSSPILSITISEDGSMSIKNEEEEQTL\n",
      "TILIRTGLLVISGLFPVSIPITAAAWYLWEVKKQR\n",
      "\n",
      ">NS3_Proteína no estructural 3\n",
      "AGVLWDVPSPPPVGRAELEDGAYRI\n",
      "KQKGILGYSQIGAGVYKEGTFHTMWHVTRGAVLMHKGKRIEPSWADVKKDLISYGGGWKL\n",
      "EGEWKEGEEVQVLALEPGKNPRAVQTKPGLFRTNTGTIGAVSLDFSPGTSGSPIVDKKGK\n",
      "VVGLYGNGVVTRGGAYVSAIAQTEKGIEDNPEIEDDIFRKRRLTIMDLHPGAGKTKRYLP\n",
      "AIVREAIKRGLRTLILAPTRVVAAEMEEALRGLPIRYQTPAIRAEHTGREIVDLMCHATF\n",
      "TMRLLSPIRVPNYNLIIMDEAHFTDPASIAARGYISTRVEMGEAAGIFMTATPPGSRDPF\n",
      "PQSNAPIMDEEREIPERSWNSGHEWVTDFKGKTVWFVPSIKTGNDIAACLRKNGKRVIQL\n",
      "SRKTFDSEYVKTRTNDWDFVVTTDISEMGANFKAERVIDPRRCMKPVILTDGEERVILAG\n",
      "PMPVTHSSAAQRRGRIGRNPRNENDQYIYMGEPLENDEDCAHWKEAKMLLDNINTPEGII\n",
      "PSMFEPEREKVDAIDGEYRLRGEARKTFVDLMRRGDLPVWLAYKVAAEGINYADRRWCFD\n",
      "GTRNNQILEENVEVEIWTKEGERKKLKPRWLDARIYSDPLALKEFKEFAAGRK\n",
      "\n",
      ">NS4A_Proteína no estructural 4A\n",
      "SLTLNLI\n",
      "TEMGRLPTFMTQKARDALDNLAVLHTAEAGGKAYNHALSELPETLETLLLLTLLATVTGG\n",
      "IFLFLMSGRGIGKMTLGMCCIITASILLWYAQIQPHWIAASIILEFFLIVLLIPEPEKQR\n",
      "\n",
      "\n",
      ">NS4B_Proteína no estructural 4B\n",
      "NEMGFLEKTKKDLGLGHIATQQPESNILDIDLRPASA\n",
      "WTLYAVATTFITPMLRHSIENSSVNVSLTAIANQATVLMGLGKGWPLSKMDIGVPLLAIG\n",
      "CYSQVNPITLTAALLMLVAHYAIIGPGLQAKATREAQKRAAAGIMKNPTVDGITVIDLDP\n",
      "IPYDPKFEKQLGQVMLLVLCVTQVLMMRTTWALCEALTLATGPVSTLWEGNPGRFWNTTI\n",
      "AVSMANIFRGSYLAGAGLLFSIMKNTTSTRR\n",
      "\n",
      ">NS5_Proteína no estructural 5\n",
      "GTGNMGETLGEKWKNRLNALGKSEFQIYK\n",
      "KSGIQEVDRTLAKEGIKRGETDHHAVSRGSAKLRWFVERNLVTPEGKVVDLGCGRGGWSY\n",
      "YCGGLKNVREVKGLTKGGPGHEEPIPMSTYGWNLVRLQSGVDVFFVPPEKCDTLLCDIGE\n",
      "SSPNPTVEAGRTLRVLNLVENWLNNNTQFCVKVLNPYMPSVIERMETLQRKYGGALVRNP\n",
      "LSRNSTHEMYWVSNASGNIVSSVNMISRMLINRFTMRHKKATYEPDVDLGSGTRNIGIES\n",
      "ETPNLDIIGKRIEKIKQEHETSWHYDQDHPYKTWAYHGSYETKQTGSASSMVNGVVRLLT\n",
      "KPWDVIPMVTQMAMTDTTPFGQQRVFKEKVDTRTQEPKEGTKKLMKITAEWLWKELGKKK\n",
      "TPRMCTREEFTKKVRSNAALGAIFTDENKWKSAREAVEDNRFWELVDKERNLHLEGKCET\n",
      "CVYNMMGKREKKLGEFGKAKGSRAIWYMWLGARFLEFEALGFLNEDHWFSRENSLSGVEG\n",
      "EGLHKLGYILREVSKKEGGAMYADDTAGWDTRITIEDLKNEEMITNHMAGEHKKLAEAIF\n",
      "KLTYQNKVVRVQRPTPRGTVMDIISRRDQRGSGQVVTYGLNTFTNMEAQLIRQMEGEGVF\n",
      "KSIQHLTASEEIAVQDWLVRVGRERLSRMAISGDDCVVKPLDDRFAKALTALNDMGKVRK\n",
      "DIQQWEPSRGWNDWTQVPFCSHHFHELIMKDGRTLVVPCRNQDELIGRARISQGAGWSLR\n",
      "ETACLGKSYAQMWSLMYFHRRDLRLAANAICSAVPSHWVPTSRTTWSIHASHEWMTTEDM\n",
      "LTVWNRVWILENPWMEDKTPVESWEEIPYLGKREDQWCGSLIGLTSRATWAKNIQTAINQ\n",
      "VRSLIGNEEYTDYMPSMKRFRREEEEVGVLW\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Recuperar, visualizar y guardar cada una de las secuencias aminoacídicas de las proteínas del DENV-2 (Formato fasta)\n",
    "result_fasta = service.search(query, frmt = 'fasta')\n",
    "\n",
    "file_secuencias = 'Uniprot_Proteínas virales_DENV-2'\n",
    "\n",
    "with open(file_secuencias, 'w') as file:\n",
    "    print('>ancC_Proteína de la cápside anclada', file = file)\n",
    "    print(result_fasta[111:212], file = file)\n",
    "    print(file = file)\n",
    "    print('>prM_Proteína premembrana', file = file)\n",
    "    print(result_fasta[226:395], file = file)\n",
    "    print(file = file)\n",
    "    print('>E_Proteína de envoltura', file = file)\n",
    "    print(result_fasta[395:898], file = file)\n",
    "    print(file = file)\n",
    "    print('>NS1_Proteína no estructural 1', file = file)\n",
    "    print(result_fasta[898:1256], file = file)\n",
    "    print(file = file)\n",
    "    print('>NS2A_Proteína no estructural 2A', file = file)\n",
    "    print(result_fasta[1256:1478], file = file)\n",
    "    print(file = file)\n",
    "    print('>NS2B_Proteína no estructural 2B', file = file)\n",
    "    print(result_fasta[1478:1610],file = file)\n",
    "    print(file = file)\n",
    "    print('>NS3_Proteína no estructural 3', file = file)\n",
    "    print(result_fasta[1610:2238], file = file)\n",
    "    print(file = file)\n",
    "    print('>NS4A_Proteína no estructural 4A', file = file)\n",
    "    print(result_fasta[2238:2368], file = file)\n",
    "    print(file = file)\n",
    "    print('>NS4B_Proteína no estructural 4B', file = file)\n",
    "    print(result_fasta[2391:2643], file = file)\n",
    "    print(file = file)\n",
    "    print('>NS5_Proteína no estructural 5', file = file)\n",
    "    print(result_fasta[2643:3558], file = file)\n",
    "    \n",
    "with open('Uniprot_Proteínas virales_DENV-2') as f:\n",
    "    contents = f.read()\n",
    "    print(contents)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
